src/code/z_rcp.c

```diff
@@ -847,28 +847,50 @@ Gfx gEmptyDL[] = {
     gsSPEndDisplayList(),
 };
 
+/**
+ * Set fog color and range.
+ *
+ * At or prior to fog near, geometry is unaffected by fog. At or beyond fog far, geometry is fully fogged.
+ * Between near and far pixels will be linearly interpolated between the unfogged color and the supplied fog color.
+ *
+ * Fog far should be in the range 0 to 1000 and greater than or equal to fog near. If fog near is negative everything
+ * will be fully fogged. If fog near is 1000 or greater there is no fog.
+ */
 Gfx* Gfx_SetFog(Gfx* gfx, s32 r, s32 g, s32 b, s32 a, s32 near, s32 far) {
+    // Ensure fog far is greater than near
     if (far == near) {
         far++;
     }
-
     ASSERT(near != far, "n != f", "../z_rcp.c", 1155);
 
+    // Set the fog color, far away geometry will be rendered as this solid color.
     gDPSetFogColor(gfx++, r, g, b, a);
 
     if (near >= 1000) {
+        // Fog far is expected to be at most 1000 so near >= far. Set a constant shade alpha of 0 for no fog
         gSPFogFactor(gfx++, 0, 0);
-    } else if (near >= 997) {
-        gSPFogFactor(gfx++, 0x7FFF, 0x8100);
+    } else if (near > 996) {
+        // Calculating the fog multiplier when far = 1000 and near >= 997 using `128000 / (far - near)` as in
+        // SPFogPosition would result in an overflow since 128000 / (1000 - 997) > 32767. Avoid this overflow by
+        // effectively clamping near to ~996. This is almost SPFogPosition(996.0937f, 1000)
+        gSPFogFactor(gfx++, 0x7FFF, -0x7F00); // Fixed point: 0x7FFF = 1.0, -0x7F00 = -0.992
```

Comment:
fm and fo are s0.15 ? Where is this from
Seems inconsistent with the formula
`fogAlpha=fm*z+fo   z in [-1;1]  fogAlpha in [0;255]`
(cf gbi.h)

Reply:
`f(z) = clamp(32767 * z - 32512, 0, 255)` for z in [-1,1] produces the curve (only the interesting part shown)

![image](https://github.com/zeldaret/oot/assets/17233964/1eca1628-869f-440f-9820-df37879db37e)

32512 is chosen so that f(z)=255 at z=1

There's some weirdness in how the microcode actually computes this since z itself isn't really in the interval [-1,1], it's in s15.16 format. The calculation is something like
```
vmudh $v__, fo, [1,1,1,1,1,1,1,1] // ACC = (fo * [1,1,1,1,1,1,1,1]) << 16
vmadh $v__, [0, 0, 0, 1, 0, 0, 0, 1], v31_0x7F00 // ACC += ([0,0,0,1,0,0,0,1] * 0x7F00) << 16
vmadn $v26, pos, fm // frac. part, $v26 = low16(ACC += pos * fm)
vmadh $v25, pos, fm // int. part, $v25 = mid16(ACC += (pos * fm) << 16)
```
https://github.com/Mr-Wiseguy/f3dex2/blob/master/f3dex2.s#L1432
I have yet to take the time to carefully check the units involved here

Reply:
then if to interpret `0x7FFF` and the other as s0.15 you rely on interpreting z as s15.16 that should be written somewhere too, otherwise it's kinda arbitrary/useless to use any unit

Reply:
looking at this again I don't understand how `fm*z+fo` with z as s15.16 and fm, fo as s0.15 produces a meaningful alpha in range 0-255
I'll remove the comment for now
```suggestion
        gSPFogFactor(gfx++, 0x7FFF, -0x7F00);
```